### ISSUE TITLE
Background Task Authentication

### ISSUE THREAD

**Issue by MarcorOnline** [OneDrive/onedrive-sdk-csharp#44]
Hello.
I'm writing a Windows 10 UWP and I want to authenticate the user into background task to sync some small files.

In foreground I do this and it works:
var oneDriveClient = await OneDriveClientExtensions.GetAuthenticatedClientUsingOnlineIdAuthenticator(scopes);

If I try to do it inside the Background task it doesn't works.

Thanks for your help

**Comment by ginach** [comment]
The OnlineIdAuthenticator requires user interaction to authenticate the user, making it not functional within a background task.

Unfortunately, Microsoft Account doesn't have an OAuth flow that never requires user interaction. Our recommended approach for apps that require this is to prompt the user for initial authentication, including the wl.offline_access scope, and store the refresh tokens for the users to use in future authentication. This could be done with a sign up page, for example. You can then use this to retrieve an authenticated client using the refresh token:

```csharp
await OneDriveClient.GetSilentlyAuthenticatedMicrosoftAccountClient(
    clientId,
    returnUrl,
    scopes,
    refreshToken)
```
A new refresh token is issued every time you authenticate with the old one and each token is good for up to a year.

**Comment by MarcorOnline** [comment]
Your solution is perfect for my app, but I don't know what I've to use as "returnUrl"

I tried this but it doesn't works:

var oneDriveClient = await OneDriveClient.GetSilentlyAuthenticatedMicrosoftAccountClient(clientId, "https://login.live.com/oauth20_desktop.srf", scopes, token);

Thanks for your help!

**Comment by ericleigh007** [comment]
+1 for not knowing how to get the refresh token in the first place.  The only path I see that returns them is the `GetMicrosoftAccountClient()` path to `GetAuthorizationCodeAsync()`, but there isn't a complete example of that path.  A working code snippet would help if you can find the time.

-thanks a bunch
-e

**Comment by ginach** [comment]
[OnlineIdAuthenticator](https://msdn.microsoft.com/en-us/library/windows/apps/windows.security.authentication.onlineid.onlineidauthenticator.aspx) doesn't return refresh token. It's an API that's specific to MSA and handles all of the OAuth logic behind the scenes, including refreshing tokens without user prompt when possible. Unfortunately, if you want to do your own storing of these tokens for use offline it's not the right authentication API to use. Instead, you'll want to use [WebAuthenticationBroker](https://msdn.microsoft.com/en-us/library/windows/apps/windows.security.authentication.web.webauthenticationbroker.aspx). The extension is OneDriveClientExtensions.GetClientUsingWebAuthenticationBroker.

It looks like there's currently a bug in the WebAuthenticationBroker flow where it is using the token OAuth flow, not code, which doesn't support returning the refresh token. I should have a PR with a fix out by early tomorrow.

**Comment by ericleigh007** [comment]
Thanks Gina!
I thought something was up, but wasn't sure what it was.  I was already using `WebAuthenticationBroker`, so we're good there.
I will work on other things in the meantime.
You saved me more time troubleshooting down a rabbit hole.

Cheers,
-e

**Comment by ginach** [comment]
@ericleigh007, my current PR into master has a fix for this if you want to take a look.

**Comment by ericleigh007** [comment]
I saw the PR yesterday but thought I would wait until it was merged.
I'll take a look now, thx
-e

**Comment by ginach** [comment]
This is fixed in version 1.1.29. Please re-open if you're still seeing issues.

**Comment by ericleigh007** [comment]
Works as advertised.  Can login that way and the refresh token is being returned now.

The only kink was having to now mark my app as "mobile" [here] (https://account.live.com/developers/applications/index) because `WebAuthenticationBroker.GetCurrentApplicationCallbackUri` returned the ms-app: Uri, and then the Server complained about the return_url not matching.

Thanks again.
-e
